zipkin-server/src/main/java/zipkin2/server/internal/ZipkinQueryApiV2.java

```diff
@@ -71,116 +64,109 @@
     this.autocompleteKeys = autocompleteKeys;
   }
 
-  @RequestMapping(
-      value = "/dependencies",
-      method = RequestMethod.GET,
-      produces = APPLICATION_JSON_VALUE)
-  public byte[] getDependencies(
-      @RequestParam(value = "endTs", required = true) long endTs,
-      @Nullable @RequestParam(value = "lookback", required = false) Long lookback)
-      throws IOException {
+  @Get("/api/v2/dependencies")
+  public HttpResponse getDependencies(
+    @Param("endTs") long endTs,
+    @Param("lookback") Optional<Long> lookback) throws IOException {
     Call<List<DependencyLink>> call =
-        storage.spanStore().getDependencies(endTs, lookback != null ? lookback : defaultLookback);
-    return DependencyLinkBytesEncoder.JSON_V1.encodeList(call.execute());
+      storage.spanStore().getDependencies(endTs, lookback.orElse(defaultLookback));
+    return jsonResponse(DependencyLinkBytesEncoder.JSON_V1.encodeList(call.execute()));
   }
 
-  @RequestMapping(value = "/services", method = RequestMethod.GET)
-  public ResponseEntity<List<String>> getServiceNames() throws IOException {
+  @Get("/api/v2/services")
+  public HttpResponse getServiceNames() throws IOException {
     List<String> serviceNames = storage.spanStore().getServiceNames().execute();
     serviceCount = serviceNames.size();
-    return maybeCacheNames(serviceNames);
+    return maybeCacheNames(serviceCount > 3, serviceNames);
   }
 
-  @RequestMapping(value = "/spans", method = RequestMethod.GET)
-  public ResponseEntity<List<String>> getSpanNames(
-      @RequestParam(value = "serviceName") String serviceName) throws IOException {
-    return maybeCacheNames(storage.spanStore().getSpanNames(serviceName).execute());
+  @Get("/api/v2/spans")
+  public HttpResponse getSpanNames(@Param("serviceName") String serviceName) throws IOException {
+    List<String> spanNames = storage.spanStore().getSpanNames(serviceName).execute();
+    return maybeCacheNames(serviceCount > 3, spanNames);
   }
 
-  @RequestMapping(value = "/traces", method = RequestMethod.GET, produces = APPLICATION_JSON_VALUE)
-  public String getTraces(
-      @Nullable @RequestParam(value = "serviceName", required = false) String serviceName,
-      @Nullable @RequestParam(value = "spanName", required = false) String spanName,
-      @Nullable @RequestParam(value = "annotationQuery", required = false) String annotationQuery,
-      @Nullable @RequestParam(value = "minDuration", required = false) Long minDuration,
-      @Nullable @RequestParam(value = "maxDuration", required = false) Long maxDuration,
-      @Nullable @RequestParam(value = "endTs", required = false) Long endTs,
-      @Nullable @RequestParam(value = "lookback", required = false) Long lookback,
-      @RequestParam(value = "limit", defaultValue = "10") int limit)
-      throws IOException {
+  @Get("/api/v2/traces")
+  public HttpResponse getTraces(
+    @Param("serviceName") Optional<String> serviceName,
```

Comment:
You may know but you can omit the value of `@Param` annotation if you specify `-parameters` javac option.
https://line.github.io/armeria/server-annotated-service.html#parameter-injection